url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/resid/EnvelopeGenerator.java

```diff
@@ -0,0 +1,518 @@
+   * <P>
+   * From the sustain levels it follows that both the low and high 4 bits of
+   * the envelope counter are compared to the 4-bit sustain value. This has
+   * been verified by sampling ENV3.
+   */
+  protected static int /* reg8 */sustain_level[] = {
+    0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa,
+    0xbb, 0xcc, 0xdd, 0xee, 0xff, };
+
+  // ----------------------------------------------------------------------------
+  // Inline functions.
+  // The following functions are defined inline because they are called every
+  // time a sample is calculated.
+  // ----------------------------------------------------------------------------
+
+  /**
+   * SID clocking - 1 cycle.
+   */
+  public void clock() {
+    // Check for ADSR delay bug.
+    // If the rate counter comparison value is set below the current value
+    // of the rate counter, the counter will continue counting up until it
+    // wraps
+    // around to zero at 2^15 = 0x8000, and then count rate_period - 1
+    // before the
+    // envelope can finally be stepped.
+    // This has been verified by sampling ENV3.
+    //
+    if ((++rate_counter & 0x8000) != 0) {
+      ++rate_counter;
+      rate_counter &= 0x7fff;
+    }
+
+    if (rate_counter != rate_period) {
+      return;
+    }
+
+    rate_counter = 0;
+
+    // The first envelope step in the attack state also resets the
+    // exponential counter. This has been verified by sampling ENV3.
+    //
+    if (state == State.ATTACK
+        || ++exponential_counter == exponential_counter_period) {
+      exponential_counter = 0;
+
+      // Check whether the envelope counter is frozen at zero.
+      if (hold_zero) {
+        return;
+      }
+
+      if (state == State.ATTACK) {
+        // The envelope counter can flip from 0xff to 0x00 by changing
+        // state to release, then to attack. The envelope counter is
+        // then frozen
+        // at zero; to unlock this situation the state must be changed
+        // to
+        // release, then to attack. This has been verified by sampling
+        // ENV3.
+        //
+        ++envelope_counter;
+        envelope_counter &= 0xff;
+        if (envelope_counter == 0xff) {
+          state = State.DECAY_SUSTAIN;
+          rate_period = rate_counter_period[decay];
+        }
+      } else if (state == State.DECAY_SUSTAIN) {
+        if (envelope_counter != sustain_level[sustain]) {
+          --envelope_counter;
+        }
+      } else if (state == State.RELEASE) {
+        // The envelope counter can flip from 0x00 to 0xff by changing
+        // state to
+        // attack, then to release. The envelope counter will then
+        // continue
+        // counting down in the release state.
+        // This has been verified by sampling ENV3.
+        // NB! The operation below requires two's complement integer.
+        //
+        --envelope_counter;
+        envelope_counter &= 0xff;
+      }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 247:262, 358:374.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.